builddefs/common_features.mk

```diff
@@ -44,6 +44,31 @@ else ifeq ($(strip $(DEBUG_MATRIX_SCAN_RATE_ENABLE)), api)
     OPT_DEFS += -DDEBUG_MATRIX_SCAN_RATE
 endif
 
+ifeq ($(strip $(CONSOLE_ENABLE)), yes)
+    SENDCHAR_DRIVER ?= console
```

Comment:
How about:
```suggestion
    CONSOLE_DRIVER ?= hid
```
etc

Reply:
I don't think we can use the terminology "hid" as there is currently still the intention to broadcast logs over XAP, which is hid.

As for `CONSOLE_DRIVER`, i would be keen to break the assumption that "console = logging". In part the code is only as it is right now from existing conventions. If we are changing it, I would suggest something more generally used by logging frameworks. However as we already have the `_DRIVER` concept, extending this to "sync" / "transport" / "appender" might be too far?

Reply:
Then perhaps something like `pjrc` or `hid_listen`?

Reply:
> I don't think we can use the terminology "hid" as there is currently still the intention to broadcast logs over XAP, which is hid.

Wouldn't the traditional console endpoint be removed once XAP lands anyway?

Reply:
> Wouldn't the traditional console endpoint be removed once XAP lands anyway?

I guess it could be viewed as redundant, though till now ive not seen any discussion on its removal.

For me, I am still keen to not tie the terminology to "console". This would extend to refactoring of the existing print/debug api to have a `console_` prefix, and prefer the more common `log_info`, logger_debug` etc.

> However as we already have the _DRIVER concept, extending this to "sync" / "transport" / "appender" might be too far?

So here i was more thinking it would be things like,

- `LOG_DRIVER = X`
- `LOGGING_APPENDER = Y`
- `LOGGER_SINK = Z`

We could jam this in for the short term and lock down the terminology before the develop cycle ends? That way we can get XAP and #22538 working?

Reply:
> For me, I am still keen to not tie the terminology to "console". This would extend to refactoring of the existing print/debug api to have a console_ prefix, and prefer the more common log_info, logger_debug` etc.

Agreed. The whole `printing` system should move to `log_xxx` or the `pr_xxx` scheme like it is used in the linux kernel.

 
> * LOG_DRIVER = X
> * LOGGING_APPENDER = Y
> * LOGGER_SINK = Z

`LOGGER_SINK` seems correct to me, but to keep the existing terminology consistent I would prefer `LOG_DRIVER`.

> We could jam this in for the short term and lock down the terminology before the develop cycle ends?

That is fine for me.

Reply:
@fauxpark friendly ping: any objections to merging this as it is?

Reply:
Not really a fan of `SENDCHAR_DRIVER` but if it's only temporary I'm fine with that. Would prefer to be more specific with the `console` value, especially if we want to distinguish it from the XAP implementation. So perhaps something like `hid_pjrc` and eventually add `hid_xap`?

Reply:
I have no strong opinion on this - so `hid_pjrc` and `hid_xap` would be fine with me.

Reply:
Hi @fauxpark and @zvecr, I think that if we can find a consensus on the `SENDCHAR_DRIVER` naming this PR can be merged. As stated in my last comment `hid_pjrc` is fine, I expect most users to never touch the driver anyway or only change it to someting like `custom` or `rtt`.

Reply:
👍

Reply:
I do not see the value in naming it anything different.

Reply:
Alright, I guess we just leave it as it is now then.

Reply:
As I suggested above I would prefer `hid_pjrc` and `hid_xap`, and I don't really understand how there is no value in that over the much more vague `console` and (presumably) `xap`. Unless the plan is to specify the transport in a separate flag?

Reply:
The user facing feature is still "console", and that isn't going to change any time soon. Plus it is exposed in docs/tutorials/blogs/etc.

Having a different value makes the mapping less direct, and wastes effort to digest.

Reply:
I'm not talking about renaming the feature though... and what you just said has further confused me considering:

> As for `CONSOLE_DRIVER`, i would be keen to break the assumption that "console = logging"

Reply:
> I'm not talking about renaming the feature though

Which reinforces my point that we should keep the 1:1 mapping. This is just the "console" logging backend for the "console" feature. 

>  and what you just said has further confused me considering:
>> As for CONSOLE_DRIVER, i would be keen to break the assumption that "console = logging"

A user should not **have** to enable console to get something like #22538. And its terminology should not bleed into the lower part of what a logging framework provides. Thus my comment on prefixing all the logging functions with `console_` should also avoided. Where longer term, the direction is to move to something more standardised (<https://github.com/rxi/log.c> returning from a quick search, or <https://github.com/qmk/qmk_firmware/pull/22159#discussion_r1431040980>).

Reply:
Maybe I should try to clarify what my mental model of this is.
So far, the convention for a feature `x` has been an enable flag `x_ENABLE`, and if applicable, a driver flag `x_DRIVER`.

Therefore having:
```
CONSOLE_ENABLE = yes
SENDCHAR_DRIVER = console
```
breaks this convention in my eyes (but to be fair, so does the existing lack of say `console_` or `log_` prefixes on all the logging functions). It seems to imply that "Console" and "Sendchar" are separate things when I feel they are very much connected.

To be clear I don't much care what the feature is named; we can change `CONSOLE` => `LOGGING` or whatever works at some later point. In fact it probably *should* be renamed from Console as this really refers to the tool used to view the log output, rather than the "protocol" itself.

So all up we'd have something similar to this:

```makefile
LOGGING_ENABLE = yes

# For consumption using hid_listen, qmk console or Toolbox HID Console
LOGGING_DRIVER = hid_pjrc
# For consumption using XAP console, however that will look
LOGGING_DRIVER = hid_xap
# For consumption over UART
LOGGING_DRIVER = uart
# BYO sendchar
LOGGING_DRIVER = custom
# etc...
```

with eg.

```c
...

log_info("Blah");
log_debug("Debug blah");

...
```

Reply:
You are welcome to implement whatever you want.

Reply:
Sure, but I'd like to understand the rationale behind *your* implementation first. Maybe I am not getting it 🤷